bundles/org.openhab.binding.smhi/src/main/java/org/openhab/binding/smhi/internal/SmhiHandler.java

```diff
@@ -0,0 +1,421 @@
+     * If it is, set flag to indicate we have the latest forecast.
+     */
+    private void getUpdatedForecast() {
+        TimeSeries forecast;
+        ZonedDateTime referenceTime;
+        SmhiConnector apiConnection = connection;
+        if (apiConnection != null) {
+            try {
+                forecast = apiConnection.getForecast(config.latitude, config.longitude);
+            } catch (SmhiException e) {
+                logger.warn("Failed to get new forecast: {}", e.getCause().getMessage());
+                return;
+            }
+            referenceTime = forecast.getReferenceTime();
+            updateChannels(forecast);
+            if (referenceTime.isEqual(currentHour) || referenceTime.isAfter(currentHour)) {
+                hasLatestForecast = true;
+            }
+            cachedTimeSeries = forecast;
+        }
+    }
+
+    /**
+     * Get the current time rounded down to hour
+     * 
+     * @return A {@link ZonedDateTime} corresponding to the last even hour
+     */
+    private ZonedDateTime calculateCurrentHour() {
+        ZonedDateTime now = ZonedDateTime.now().withZoneSameInstant(ZoneOffset.UTC);
+        int y = now.getYear();
+        int m = now.getMonth().getValue();
+        int d = now.getDayOfMonth();
+        int h = now.getHour();
+        return ZonedDateTime.of(y, m, d, h, 0, 0, 0, ZoneOffset.UTC);
+    }
+
+    /**
+     * Get the current time rounded down to day
+     * 
+     * @return A {@link ZonedDateTime} corresponding to the last even day.
+     */
+    private ZonedDateTime calculateCurrentDay() {
+        ZonedDateTime now = ZonedDateTime.now().withZoneSameInstant(ZoneOffset.UTC);
+        int y = now.getYear();
+        int m = now.getMonth().getValue();
+        int d = now.getDayOfMonth();
+        return ZonedDateTime.of(y, m, d, 0, 0, 0, 0, ZoneOffset.UTC);
+    }
+
+    /**
+     * Creates channels based on selections in thing configuration
+     * 
+     * @return
+     */
+    private List<Channel> createChannels() {
+        List<Channel> channels = new ArrayList<>();
+
+        // There's currently a bug in PaperUI that can cause options to be added more than one time
+        // to the list. Convert to a sorted set to work around this.
+        // See https://github.com/openhab/openhab-webui/issues/212
+        Set<Integer> hours = new TreeSet<>();
+        Set<Integer> days = new TreeSet<>();
+        if (config.hourlyForecasts != null) {
+            hours.addAll(config.hourlyForecasts);
+        }
+        if (config.dailyForecasts != null) {
+            days.addAll(config.dailyForecasts);
+        }
+
+        for (int i : hours) {
+            ChannelGroupUID groupUID = new ChannelGroupUID(thing.getUID(), "hour_" + i);
+            CHANNEL_IDS.forEach(id -> {
+                ChannelUID channelUID = new ChannelUID(groupUID, id);
+                Channel channel = ChannelBuilder.create(channelUID, "Number")
```

Comment:
Don't most of these channels have units now?

Reply:
The units are working even though it's defined just as "Number". Do I have to hardcode all Item types here as well? Is there an easier way to just create a channel from a channel-type defined in xml? All info is already defined there, so seems unnecessary to define it again here, but the ChannelBuilder constructor requires this.

Reply:
Not sure myself. But after looking at the core code it seems like you can just set the itemType to null. If the core was coded sensibly it would look at the ChannelType's itemType in the case that the Channel's itemType was missing. I think that would be one way to avoid redefining the itemtypes if it works.

Reply:
Thanks, will try and test it out!

Reply:
Didn't work. The acceptedItemType property remained null, giving no hint to the UI what item type to use. Had to do it the harder way instead.